bundles/org.openhab.binding.iaqualink/src/main/java/org/openhab/binding/iaqualink/internal/api/IAqualinkClient.java

```diff
@@ -0,0 +1,448 @@
+            if (response.getStatus() == HttpStatus.UNAUTHORIZED_401) {
+                throw new NotAuthorizedException(response.getReason());
+            }
+            if (response.getStatus() != HttpStatus.OK_200) {
+                throw new IOException(response.getReason());
+            }
+            return gson.fromJson(response.getContentAsString(), typeOfT);
+        } catch (TimeoutException | ExecutionException | JsonParseException e) {
+            throw new IOException(e);
+        }
+    }
+
+    /**
+     *
+     * @param url
+     * @return
+     * @throws IOException
+     * @throws NotAuthorizedException
+     */
+    private JsonElement getAqualinkJsonObject(String url)
+            throws InterruptedException, IOException, NotAuthorizedException {
+        try {
+            logger.trace("Trying {}", url);
+            ContentResponse response = httpClient.newRequest(url).method(HttpMethod.GET) //
+                    .agent(HEADER_AGENT) //
+                    .header(HttpHeader.ACCEPT_LANGUAGE, HEADER_ACCEPT_LANGUAGE) //
+                    .header(HttpHeader.ACCEPT_ENCODING, HEADER_ACCEPT_ENCODING) //
+                    .header(HttpHeader.ACCEPT, HEADER_ACCEPT) //
+                    .send();
+            logger.trace("Response {}", response.toString());
+            if (response.getStatus() == HttpStatus.UNAUTHORIZED_401) {
+                throw new NotAuthorizedException(response.getReason());
+            }
+            if (response.getStatus() != HttpStatus.OK_200) {
+                throw new IOException(response.getReason());
+            }
+            return parser.parse(response.getContentAsString());
+        } catch (TimeoutException | ExecutionException | JsonParseException e) {
+            throw new IOException(e);
+        }
+    }
+
+    /////////////// .........Here be dragons...../////////////////////////
+
+    class HomeDeserializer implements JsonDeserializer<Home> {
+        @Override
+        public Home deserialize(@Nullable JsonElement json, @Nullable Type typeOfT,
+                @Nullable JsonDeserializationContext context) throws JsonParseException {
+            if (json == null) {
+                throw new JsonParseException("No JSON");
+            }
+            JsonObject jsonObject = json.getAsJsonObject();
+            JsonArray homeScreen = jsonObject.getAsJsonArray("home_screen");
+            JsonObject home = new JsonObject();
+            if (homeScreen != null) {
+                homeScreen.forEach(element -> {
+                    element.getAsJsonObject().entrySet().forEach(entry -> {
+                        home.add(entry.getKey(), entry.getValue());
+                    });
+                });
+                return gsonInternal.fromJson(home, Home.class);
+            }
+            throw new JsonParseException("Invalid structure for Home class");
+        }
+    }
+
+    class OneTouchDeserializer implements JsonDeserializer<OneTouch[]> {
```

Comment:
Is it so special that it can't map this out-of-the box?

Reply:
I mention this in the class java doc, but yes the returning json is very awful, i think i also mention tongue and cheek that "here be dragons" above this code to indicate the awfulness of the returning JSON.